### PR TITLE
AP_PiccoloCAN: Fix temperature logging bug

### PR DESCRIPTION
**Pull request by SchrodingersGat**
- Fixes issues with temperature data in telemetry and logs
- Use the maximum of available temperature values for the ESC (where available)

A customer reported that the temperature data reported from a [Velocity ESC](https://www.currawongeng.com/velocity-esc/) was incorrect. It appears that the available temperature readings are not being directed into the logging structures in the correct fashion.

A pretty simple change to fix the issue.

Another small change is to choose the "maximum" ESC temperature (for variants of the ESC which have multiple temperature readings available).

### CODE REVIEW

File: libraries/AP_PiccoloCAN/AP_PiccoloCAN.cpp [hunk 1]
```diff
@@ -376,7 +376,8 @@ void AP_PiccoloCAN::send_esc_telemetry_mavlink(uint8_t mav_chan)
         if (is_esc_present(ii)) {
             dataAvailable = true;
 
-            temperature[idx] = esc.fetTemperature;
+            // Provide the maximum ESC temperature in the telemetry stream
+            temperature[idx] = esc.fetTemperature > esc.escTemperature ? esc.fetTemperature : esc.escTemperature;
```
Comment (peterbarker) [inline libraries/AP_PiccoloCAN/AP_PiccoloCAN.cpp:380]:
```suggestion
            temperature[idx] = MAX(esc.fetTemperature, esc.escTemperature);
```

File: libraries/AP_PiccoloCAN/AP_PiccoloCAN.cpp [hunk 2]
```diff
@@ -691,21 +692,27 @@ bool AP_PiccoloCAN::handle_esc_message(AP_HAL::CANFrame &frame)
     } else if (decodeESC_StatusBPacket(&frame, &esc.voltage, &esc.current, &esc.dutyCycle, &esc.escTemperature, &esc.motorTemperature)) {
         
         AP_ESC_Telem_Backend::TelemetryData t {
-            .temperature_cdeg = int16_t(esc.escTemperature * 100),
             .voltage = float(esc.voltage) * 0.01f,
             .current = float(esc.current) * 0.01f,
+            .motor_temp_cdeg = int16_t(esc.motorTemperature * 100),
```
Comment (peterbarker) [inline libraries/AP_PiccoloCAN/AP_PiccoloCAN.cpp:697]: Can't rearrange fields.  It's important our code compiles :-)

File: libraries/AP_PiccoloCAN/AP_PiccoloCAN.cpp [hunk 3]
```diff
@@ -691,21 +692,27 @@ bool AP_PiccoloCAN::handle_esc_message(AP_HAL::CANFrame &frame)
...
         };
 
         update_telem_data(addr, t,
             AP_ESC_Telem_Backend::TelemetryType::CURRENT
-                | AP_ESC_Telem_Backend::TelemetryType::VOLTAGE
-                | AP_ESC_Telem_Backend::TelemetryType::TEMPERATURE);
+            | AP_ESC_Telem_Backend::TelemetryType::VOLTAGE
+            | AP_ESC_Telem_Backend::TelemetryType::MOTOR_TEMPERATURE);
```
Comment (peterbarker) [inline libraries/AP_PiccoloCAN/AP_PiccoloCAN.cpp:703]: Bad whitespace change

File: libraries/AP_PiccoloCAN/AP_PiccoloCAN.cpp [hunk 4]
```diff
@@ -691,21 +692,27 @@ bool AP_PiccoloCAN::handle_esc_message(AP_HAL::CANFrame &frame)
...
 
         esc.newTelemetry = true;
     } else if (decodeESC_StatusCPacket(&frame, &esc.fetTemperature, &esc.pwmFrequency, &esc.timingAdvance)) {
-        AP_ESC_Telem_Backend::TelemetryData t { };
-        t.motor_temp_cdeg = int16_t(esc.fetTemperature * 100);
-        update_telem_data(addr, t, AP_ESC_Telem_Backend::TelemetryType::MOTOR_TEMPERATURE);
+
+        // Use the higher reported value of 'escTemperature' and 'fetTemperature'
+        int16_t escTemp = esc.fetTemperature > esc.escTemperature ? esc.fetTemperature : esc.escTemperature;
```
Comment (peterbarker) [inline libraries/AP_PiccoloCAN/AP_PiccoloCAN.cpp:709]:
```suggestion
        const int16_t escTemp = MAX(esc.fetTemperature, esc.escTemperature);
```

File: libraries/AP_PiccoloCAN/AP_PiccoloCAN.cpp [hunk 5]
```diff
@@ -691,21 +692,27 @@ bool AP_PiccoloCAN::handle_esc_message(AP_HAL::CANFrame &frame)
...
+
+        AP_ESC_Telem_Backend::TelemetryData t {
```
Comment (peterbarker) [inline libraries/AP_PiccoloCAN/AP_PiccoloCAN.cpp:711]:
```suggestion
        const AP_ESC_Telem_Backend::TelemetryData t {
```

**Review by peterbarker — CHANGES_REQUESTED** [review]
We really only want one commit for this change, too :-)

File: libraries/AP_PiccoloCAN/AP_PiccoloCAN.cpp (hunk 2 above)
Comment (SchrodingersGat) [reply libraries/AP_PiccoloCAN/AP_PiccoloCAN.cpp:697]: > We really only want one commit for this change, too :-)

Should I resubmit as a PR with a single commit? Happy to fix all the other points as per your suggestions.

**Comment by SchrodingersGat** [comment]
@peterbarker all fixed, I think. Please let me know if there are other issues!

**Comment by SchrodingersGat** [comment]
@peterbarker any further requests for this PR?

**Comment by SchrodingersGat** [comment]
@peterbarker builds fixed :)

File: libraries/AP_PiccoloCAN/AP_PiccoloCAN.cpp [hunk 6]
```diff
@@ -690,22 +691,28 @@ bool AP_PiccoloCAN::handle_esc_message(AP_HAL::CANFrame &frame)
...
-            .temperature_cdeg = int16_t(esc.escTemperature * 100),
-            .voltage = float(esc.voltage) * 0.01f,
-            .current = float(esc.current) * 0.01f,
-        };
+        AP_ESC_Telem_Backend::TelemetryData telem = {};
 
-        update_telem_data(addr, t,
+        telem.voltage = float(esc.voltage) * 0.01f;
+        telem.current = float(esc.current) * 0.01f;
+        telem.motor_temp_cdeg = int16_t(esc.motorTemperature * 100);
```
Comment (peterbarker) [inline libraries/AP_PiccoloCAN/AP_PiccoloCAN.cpp:698]: You're updating three fields - voltage, current and motor_temp_cdeg, but in the update call you're saying that CURRENT, VOLTAGE and TEMPERATURE are updated.  That means the `update_telem_data` code will be looking in the `temperature_cdeg` field for information *not* in `motor_temp_cdeg`.

If I'm right then the motor temperature reported here will always be zero.  We really should have a simulator to make sure this sort of thing continues to function at a basic level....

File: libraries/AP_PiccoloCAN/AP_PiccoloCAN.cpp [hunk 7]
```diff
@@ -690,22 +691,28 @@ bool AP_PiccoloCAN::handle_esc_message(AP_HAL::CANFrame &frame)
...
         esc.newTelemetry = true;
     } else if (decodeESC_StatusCPacket(&frame, &esc.fetTemperature, &esc.pwmFrequency, &esc.timingAdvance)) {
-        AP_ESC_Telem_Backend::TelemetryData t { };
-        t.motor_temp_cdeg = int16_t(esc.fetTemperature * 100);
-        update_telem_data(addr, t, AP_ESC_Telem_Backend::TelemetryType::MOTOR_TEMPERATURE);
+
+        // Use the higher reported value of 'escTemperature' and 'fetTemperature'
+        const int16_t escTemp = MAX(esc.fetTemperature, esc.escTemperature);
+
+        AP_ESC_Telem_Backend::TelemetryData telem = {};
```
Comment (peterbarker) [inline libraries/AP_PiccoloCAN/AP_PiccoloCAN.cpp:711]:
```suggestion
        AP_ESC_Telem_Backend::TelemetryData telem {};
```

.... usually the compiler will elide this sort of thing - but if it doesn't what you end up doing is contructing an `AP_ESC_Telem_Backend::TelemetryData` on the stack and then copying it back over the `telem` `AP_ESC_Telem_Backend::TelemetryData`.  Omitting the ` =` avoids this.

File: libraries/AP_PiccoloCAN/AP_PiccoloCAN.cpp [hunk 8]
```diff
@@ -690,22 +691,28 @@ bool AP_PiccoloCAN::handle_esc_message(AP_HAL::CANFrame &frame)
...
 
-        update_telem_data(addr, t,
+        telem.voltage = float(esc.voltage) * 0.01f;
+        telem.current = float(esc.current) * 0.01f;
+        telem.motor_temp_cdeg = int16_t(esc.motorTemperature * 100);
+
+        update_telem_data(addr, telem,
             AP_ESC_Telem_Backend::TelemetryType::CURRENT
                 | AP_ESC_Telem_Backend::TelemetryType::VOLTAGE
                 | AP_ESC_Telem_Backend::TelemetryType::TEMPERATURE);
```
Comment (peterbarker) [inline libraries/AP_PiccoloCAN/AP_PiccoloCAN.cpp:703]:
```suggestion
                | AP_ESC_Telem_Backend::TelemetryType::MOTOR_TEMPERATURE);
```

**Review by peterbarker — CHANGES_REQUESTED** [review]
Please check that the motor temperature is reported correctly on the GCS....

File: libraries/AP_PiccoloCAN/AP_PiccoloCAN.cpp (hunk 8 above)
Comment (SchrodingersGat) [reply libraries/AP_PiccoloCAN/AP_PiccoloCAN.cpp]: Good catch @peterbarker!

**Review by peterbarker — APPROVED** [review]
LGTM

**Review by tridge — APPROVED** [review]

**Pull request merged**